test/allDocsGenerator.js

```diff
@@ -0,0 +1,140 @@
+            assert.deepStrictEqual(actualDoc, expectedDocs.rows[key].doc);
+          }
+          assert.ok(Number.isInteger(chunk.batch));
+          assert.ok(Number.isInteger(chunk.total));
+          if (chunk.batch === 3) {
+            assert.strictEqual(chunk.docs.length, 2); // smaller last batch
+          } else {
+            assert.strictEqual(chunk.docs.length, 3);
+          }
+          actualTotal = chunk.total;
+          iter++;
+          callback();
+        }
+      })
+    );
+
+    assert.strictEqual(actualTotal, 11);
+    assert.ok(couch.isDone());
+  });
+
+  it('should return expected items with transient error', async function() {
+    let actualTotal = 0;
+    // Query string keys are stringified by Nano
+    const badgerKey = 'badger\0';
+    const kookaburraKey = 'kookaburra\0';
+    const snipeKey = 'snipe\0';
+    let iter = 1;
+
+    const couch = nock(dbUrl)
+      // batch 1
+      .post('/_all_docs', { limit: 3, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_1.json', { 'Content-Type': 'application/json' })
+      // batch 2
+      .post('/_all_docs', { limit: 3, start_key: badgerKey, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_2.json', { 'Content-Type': 'application/json' })
+      // batch 3 - transient error
+      .post('/_all_docs', { limit: 3, start_key: kookaburraKey, include_docs: true })
+      .reply(500, { error: 'Internal Server Error' })
+      // batch 3 - retry
+      .post('/_all_docs', { limit: 3, start_key: kookaburraKey, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_3.json', { 'Content-Type': 'application/json' })
+      // batch 4
+      .post('/_all_docs', { limit: 3, start_key: snipeKey, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_4.json', { 'Content-Type': 'application/json' });
+
+    await pipeline(
+      allDocsGen(dbUrl, { parallelism: 1, bufferSize: 3 }),
+      new Writable({
+        objectMode: true,
+        write: (chunk, encoding, callback) => {
+          for (const [key, actualDoc] of Object.entries(chunk.docs)) {
+            const expectedDocs = JSON.parse(fs.readFileSync(`./test/fixtures/animaldb_all_docs_${iter}.json`, 'utf8'));
+            assert.deepStrictEqual(actualDoc, expectedDocs.rows[key].doc);
+          }
+          assert.ok(Number.isInteger(chunk.batch));
+          assert.ok(Number.isInteger(chunk.total));
+          if (chunk.batch === 3) {
+            assert.strictEqual(chunk.docs.length, 2); // smaller last batch
+          } else {
+            assert.strictEqual(chunk.docs.length, 3);
+          }
+          actualTotal = chunk.total;
+          iter++;
+          callback();
+        }
+      })
+    );
+
+    assert.strictEqual(actualTotal, 11);
+    assert.ok(couch.isDone());
+  });
+});
```

Comment:
I think it would be good to have an additional test where the last batch is empty to make sure that is working.
This is also an example of asserting outside the `write`.
e.g.
```js
  it('should return expected items with empty page', async function() {
    let actualTotal = 0;
    const mockDoc = {
      _id: 'doc1',
      _rev: '1-abc',
      foo: 'bar'
    };
    const expected = [{ command: 'd', batch: 0, docs: [mockDoc] }, { command: 'd', batch: 1, docs: [] }];
    const output = [];

    const couch = nock(dbUrl)
      // first batch (0)
      .post('/_all_docs', { limit: 1, include_docs: true })
      .reply(200, {
        total_rows: 1,
        offset: 0,
        rows: [
          {
            id: 'doc1',
            key: 'doc1',
            value: {
              rev: '1-abc'
            },
            doc: mockDoc
          }
        ]
      })
      // second batch (1) (empty)
      .post('/_all_docs', { limit: 1, start_key: 'doc1\0', include_docs: true })
      .reply(200, {
        total_rows: 1,
        offset: 0,
        rows: []
      });

    await pipeline(
      allDocsGen(dbUrl, { parallelism: 1, bufferSize: 1 }),
      new Writable({
        objectMode: true,
        write: (chunk, encoding, callback) => {
          output.push(chunk);
          actualTotal += chunk.docs.length;
          callback();
        }
      })
    );

    assert.strictEqual(actualTotal, 1);
    assert.deepStrictEqual(output, expected);
    assert.ok(couch.isDone());
  });
```